BUG: Fix compile-time test of POPCNT

cmichal2 opened the pull request:
The compile-time test of POPCNT, cpu_popcnt.c produced code that would
execute without error even if the machine didn't support the popcnt
instruction. This patch changes the test so that it attempts to use popcnt on
random numbers so the compiler can't substitute the answer at compile
time. Resolves issue #19067. 

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

cmichal2 committed:
Fix compile-time test of POPCNT

The compile-time test of POPCNT, cpu_popcnt.c produced code that would
execute without error even if the machine didn't support the popcnt
instruction. This patch attempts to use popcnt on random numbers so the
compiler can't substitute the answer at compile time.

seiko2plus reviewed (changes requested):
I believe that the proposed fix will not resolve the issue. But on the contrary, it creates a new bug due to the fact that the function `random()` is not available by the compiler in the case of activating c99, which leads to the failure of the test, whether the `popcnt` is available or not.
However, the main problem is not because the compiler void generating `popcnt` instructions due to the static analysis but because the functionality
`__builtin_popcount` is always available even when `popcnt` is not enabled via `-mpopcnt` or through native flag `-march=native`.  I think we should test `_mm_popcnt_u64/_mm_popcnt_u32` on GCC instead same as we do with MSVC which will guarantee raise `target specific option mismatch` error when `popcnt` isn't supported by the host when `-march=native` is enabled. In other words, I suggest replacing the current test with the following:

```C
#ifdef _MSC_VER
    #include <nmmintrin.h>
#else
    #include <popcntintrin.h>
#endif

int main(int argc, char **argv)
{
    // To make sure popcnt instructions are generated
    // and been tested against the assembler
    unsigned long long a = *((unsigned long long*)argv[argc-1]);
    unsigned int b = *((unsigned int*)argv[argc-2]);

#if defined(_M_X64) || defined(__x86_64__)    
    a = _mm_popcnt_u64(a);
#endif
    b = _mm_popcnt_u32(b);
    return (int)a + b;
}

```

cmichal2 committed:
Change fix of cpu_popcnt.c to use _mm_popcnt_u64/_mm_popcnt_u32 on GCC

_builtin_popcount is always available, so the compile-time check always
succeeds.

cmichal2 commented:
That's embarrassing. You're right that my patch fails because random isn't available. I didn't realize, but it appears that the test isn't whether the code actually executes, its just 'does it compile?'

Your suggestion seems to work - compiling that fails on a machine without popcnt and succeeds on one that does. The code of this test case crashes with a segmentation fault on either case however.

cmichal2 commented:
The segfault is just because I executed it with arguments.

cmichal2 commented:
Is it worth including a run-time test in the test program to check if arguments were passed - so it executes properly no matter how its called? Something like:
```
#ifdef _MSC_VER
    #include <nmmintrin.h>
#else
    #include <popcntintrin.h>
#endif

int main(int argc, char **argv)
{
    // To make sure popcnt instructions are generated
    // and been tested against the assembler
  unsigned long long a; 
  unsigned int b;

  if (argc > 2){
    a = *((unsigned long long*)argv[argc-1]);
    b = *((unsigned int*)argv[argc-2]);
  }
  else{
    a = b = 1;
  }
#if defined(_M_X64) || defined(__x86_64__)    
    a = _mm_popcnt_u64(a);
#endif
    b = _mm_popcnt_u32(b);
    return (int)a + b;
}

```
This seems irrelevant if the test is never executed, but might save someone some pain in the future.

seiko2plus commented:
@cmichal2, No these compile-time tests to check the compiler support and it doesn't design to be executed on runtime. we have other runtime-tests that cover all enabled SIMD intrinsics located at [numpy/core/tests/test_simd.py](https://raw.githubusercontent.com/numpy/numpy/main/numpy/core/tests/test_simd.py).

seiko2plus reviewed (approved):
LGTM, Thank you!

mattip commented:
Thanks @cmichal2
